clients/adapter/src/main/java/org/eclipse/hono/adapter/client/command/CommandRouterClient.java

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.adapter.client.command;
+
+import java.time.Duration;
+
+import org.eclipse.hono.util.Lifecycle;
+
+import io.opentracing.SpanContext;
+import io.vertx.core.Future;
+
+/**
+ * A client for accessing Hono's Command Router API.
+ * <p>
+ * See Hono's <a href="https://www.eclipse.org/hono/docs/api/command-router">
+ * Command Router API</a> for a description of the status codes returned.
+ */
+public interface CommandRouterClient extends Lifecycle {
+
+    /**
+     * Sets the given gateway as the last gateway that acted on behalf of the given device.
+     * <p>
+     * If a device connects directly instead of through a gateway, the device identifier itself is to be used as value
```

Comment:
just out of curiosity: when would an adapter need to invoke this method with gatewayId = deviceId?

Reply:
Whether there really is a real *need* is a good question.

When a device has connected directly to a protocol adapter and has subscribed for commands there, incoming commands will always be routed to that device directly, not taking any gateways into account and not using the `lastKnownGateway` information.
The `lastKnownGateway` information is currently only used, if there was no explicit subscription for the device id of the command message, so that the `via` gateways of the device are checked, using the `lastKnownGateway` information to choose from potentially multiple subscribed gateways.

In that sense there is no need to invoke `lastKnownGateway` with gatewayId = deviceId with respect to the command routing algorithm.

Independent of that, I think the original idea of the `lastKnownGateway` information was just that it should reflect the most up-to-date information about the `device connection`, whether it be a direct connection or via a gateway, in so far as can be deduced from telemetry/event messages received at the adapter.

If the `lastKnownGateway` information continues to not be used for purposes other than command routing, we might as well skip setting it for cases with gatewayId = deviceId. But I think this would probably not have that many performance benefits over all, since a device having non-empty `via` entries and connecting directly to an adapter is probably not the usual case.